magma/circuit.py

```diff
@@ -216,11 +215,11 @@ def _get_interface_type(cls, add_default_clock):
 
 
 def _setup_interface(cls):
-    # Called from DefinitionContext.finalize.
     IO = _get_interface_type(cls, add_default_clock=True)
-    if IO is None:
+    if not hasattr(cls, "IO") and not hasattr(cls, "io"):
```

Comment:
why this check instead of `IO is None`? seems like duplicated logic from `_get_interface_type`.

Reply:
In line 206, we return None because cl.IO is already an InterfaceKind.  This happens when you do a subclass of a Circuit that already has an IO defined (so _get_interface_type was already run in the parent class).  However, we still want lines 223/224 to run for this subclass (we need to create a new `interface` attribute with `defn=cls`, otherwise we inherit the parent interface attribute with `defn=parent_cls`, we also need to update the port attribute or else the refer to the parent class).

Reply:
got it. think this logic can prob be cleaned up but looks correct.